test/functional/eval/input_spec.lua

```diff
@@ -1,38 +1,334 @@
+      return 'TEST'
+    endfunction
+    function CustomListCompl(...)
+      return ['FOO']
+    endfunction
+  ]])
+  screen:set_default_attr_ids({
+    EOB={bold = true, foreground = Screen.colors.Blue1},
+    T={foreground=Screen.colors.Red},
+  })
 end)
 
 describe('input()', function()
   it('works correctly with multiline prompts', function()
     feed([[:call input("Test\nFoo")<CR>]])
     screen:expect([[
-      {1:~                        }|
-      {1:~                        }|
-      {1:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
       Test                     |
       Foo^                      |
-    ]], {{bold=true, foreground=Screen.colors.Blue}})
+    ]])
   end)
   it('works correctly with multiline prompts and :echohl', function()
-    command('hi Test ctermfg=Red guifg=Red term=bold')
     feed([[:echohl Test | call input("Test\nFoo")<CR>]])
     screen:expect([[
-      {1:~                        }|
-      {1:~                        }|
-      {1:~                        }|
-      {2:Test}                     |
-      {2:Foo}^                      |
-    ]], {{bold=true, foreground=Screen.colors.Blue}, {foreground=Screen.colors.Red}})
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {T:Test}                     |
+      {T:Foo}^                      |
+    ]])
+    wait()
+    command('redraw!')
+    wait()
+    screen:expect([[
+                               |
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {T:Foo}^                      |
+    ]])
+  end)
+  it('works correctly with multiple numeric arguments (many args)', function()
+    command('hi Test ctermfg=Red guifg=Red term=bold')
+    feed([[:echohl Test | call input(1, 2)<CR>]])
+    wait()  -- Without wait() it first shows `12` line and then empty line.
+    command('redraw!')  -- Without this it shows two `12` lines.
+    wait()
+    -- None of the above problems happen when testing manually.
+    screen:expect([[
+                               |
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {T:1}2^                       |
+    ]])
+    feed('<BS>')
+    wait()
+    screen:expect([[
+                               |
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {T:1}^                        |
+    ]])
+  end)
+  it('works correctly with multiple numeric arguments (dict arg)', function()
+    feed([[:echohl Test | echo input({"prompt": 1, "default": 2, "cancelreturn": 3})<CR>]])
+    wait()  -- Without wait() it first shows `12` line and then empty line.
+    command('redraw!')  -- Without this it shows two `12` lines.
```

Comment:
hmm, testing locally without `redraw!` I get a slightly different `expect()` but nothing wrong like two `12` lines. The empty lines below the prompt is because the typed commandline does take multiple lines, I guess using `command()` would make sense if we want to avoid it.